### ISSUE TITLE
Last event does not aggregate properly

### ISSUE THREAD

**Issue by iti888** [logstash-plugins/logstash-filter-aggregate#50]
- Version: ES 5.02 with elasticsearch-jdbc-2.3.4.1, openjdk version "1.8.0_111"
- Operating System: ubuntu16.04.1, Oracle 12c on linux
- Config File (if you have sensitive info, please remove it):
input {
    jdbc {
        # Oracle jdbc connection string to Oracle database
        jdbc_connection_string => "jdbc:oracle:thin:@//host:port/sid"
        # The user we wish to execute our statement as
        jdbc_user => "xxx"
        jdbc_password=> "xxx"
        # The path to our downloaded jdbc driver
        jdbc_driver_library => "PATH/ojdbc7.jar"
        # The name of the driver class for Oracle
        jdbc_driver_class => "Java::oracle.jdbc.driver.OracleDriver"
        # our query
        statement => "select t1.id,t1.name,t1.description, t2.date_of_birth from t1 inner join t2 on t1.ID=t2.T1_ID"
    }
}
filter {
 aggregate {
             task_id => "%{id}"
             code => "
                map['id'] ||= event.get('id')
		map['name'] ||= event.get('name')
		map['description'] ||= event.get('description')
		
                map['date_of_birth'] ||= []
                map['date_of_birth'] <<= event.get('date_of_birth')
             "
             push_previous_map_as_event => true
             timeout => 5
	}
}
output {
    elasticsearch {
        index => "test"
        document_type => "aggregate"
        document_id => "%{id}"
        hosts => "host"
	}
}

- Sample Data:
DROP TABLE t1;

CREATE TABLE t1 (id NUMBER (10) NOT NULL, name VARCHAR2 (30 BYTE) NOT NULL, description VARCHAR2 (200 BYTE));

ALTER TABLE t1 ADD (
  CONSTRAINT t1_pk
  PRIMARY KEY
  (id)
  ENABLE VALIDATE);
COMMIT;

INSERT INTO t1
VALUES (1, 'name1', '1st name');

INSERT INTO t1
VALUES (2, 'name2', '2nd name');

INSERT INTO t1
VALUES (3, 'name3', '3rd name');

COMMIT;

drop table t2;
CREATE TABLE t2 (id NUMBER (10) NOT NULL, t1_id NUMBER (10) NOT NULL, date_of_birth DATE NOT NULL);

ALTER TABLE t2 ADD (
  CONSTRAINT t2_pk
  PRIMARY KEY
  (id)
  ENABLE VALIDATE);
COMMIT;

INSERT INTO t2
VALUES (1, 1, to_date('1/1/2000','mm/dd/yyyy'));
INSERT INTO t2
VALUES (2, 1, to_date('2/1/2000','mm/dd/yyyy'));
INSERT INTO t2
VALUES (3, 1, to_date('3/1/2000','mm/dd/yyyy'));
INSERT INTO t2
VALUES (4, 2, to_date('4/1/2000','mm/dd/yyyy'));
INSERT INTO t2
VALUES (5, 3, to_date('5/1/2000','mm/dd/yyyy'));
INSERT INTO t2
VALUES (6, 3, to_date('6/1/2000','mm/dd/yyyy'));
commit;
- Steps to Reproduce:
Create the table in Oracle 12C as script provide above
Create the config file as provided above as oracle-out_test.conf
run:
sudo ./bin/logstash -f oracle-out_test.conf --path.settings=/etc/logstash

Results are as follows:
curl -XGET localhost:9200/test/_search?pretty
{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "failed" : 0
  },
  "hits" : {
    "total" : 3,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test",
        "_type" : "aggregate",
        "_id" : "2",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2016-12-06T22:54:35.107Z",
          "date_of_birth" : [
            "2000-04-01T08:00:00.000Z"
          ],
          "name" : "name2",
          "@version" : "1",
          "description" : "2nd name",
          "id" : 2,
          "tags" : [ ]
        }
      },
      {
        "_index" : "test",
        "_type" : "aggregate",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2016-12-06T22:54:35.107Z",
          "date_of_birth" : [
            "2000-01-01T08:00:00.000Z",
            "2000-02-01T08:00:00.000Z",
            "2000-03-01T08:00:00.000Z"
          ],
          "name" : "name1",
          "@version" : "1",
          "description" : "1st name",
          "id" : 1,
          "tags" : [ ]
        }
      },
      {
        "_index" : "test",
        "_type" : "aggregate",
        "_id" : "3",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2016-12-06T22:54:34.993Z",
          "date_of_birth" : "2000-06-01T07:00:00.000Z",
          "name" : "name3",
          "@version" : "1",
          "description" : "3rd name",
          "id" : 3,
          "tags" : [ ]
        }
      }
    ]
  }
}

Did you see that the last doc date_of_birth is:           "date_of_birth" : "2000-06-01T07:00:00.000Z",
while the first doc is: "date_of_birth" : [
            "2000-04-01T08:00:00.000Z"
          ],

Last doc does not aggregate.

Also, why is the mapping like this:
curl -XGET localhost:9200/test/_mapping?pretty
{
  "test" : {
    "mappings" : {
      "aggregate" : {
        "properties" : {
          "@timestamp" : {
            "type" : "date"
          },
          "@version" : {
            "type" : "text",
            "fields" : {
              "keyword" : {
                "type" : "keyword",
                "ignore_above" : 256
              }
            }
          },
          "date_of_birth" : {
            "type" : "date"
          },
          "description" : {
            "type" : "text",
            "fields" : {
              "keyword" : {
                "type" : "keyword",
                "ignore_above" : 256
              }
            }
          },
          "id" : {
            "type" : "long"
          },
          "name" : {
            "type" : "text",
            "fields" : {
              "keyword" : {
                "type" : "keyword",
                "ignore_above" : 256
              }
            }
          }
        }
      }
    }
  }
}

I would expect to have "type":"nested" somewhere. Did I do anything wrong?

Thanks

**Comment by fbaligand** [comment]
Hi,

What is your Logstash version ? What is your logstash-filter-aggregate version ?

**Comment by fbaligand** [comment]
First, I notice one weird thing in your elasticsearch documents :
For document which id is 3, "date_of_birth" is a simple string, not a string array (contrary to 2 others).
It seems that indexed document is not the aggregated one, but the original one generated by jdbc input plugin.

Then, I tried to reproduce your problem with logstash 5.0.0 and logstash-filter-aggregate 2.5.0.
And it works nicely for me !

I make few changes to you example :
- I use a simple tcp input with json codec (more easy to me to reproduce)
- I use a stdout with ruby debug codec (more easy to me to reproduce)
- I use the `timeout_tags` aggregate feature, to easily drop not aggregated logstash events

So here is my logstash configuration :
``` ruby
input {
	tcp {
		port => 12345
		codec => json
	}
}

filter {
	aggregate {
		task_id => "%{id}"
		code => "
			map['id'] ||= event.get('id')
			map['name'] ||= event.get('name')
			map['description'] ||= event.get('description')
			map['date_of_birth'] ||= []
			map['date_of_birth'] <<= event.get('date_of_birth')
		"
		push_previous_map_as_event => true
		timeout => 5
		timeout_tags => ["aggregate"]
	}
	
	if "aggregate" not in [tags] {
		drop{}
	}
}

output {
	stdout {
		codec => rubydebug
	}
}
```

Here are my input data :
``` json
{"id" : "1", "date_of_birth" : "2000-01-01"}
{"id" : "1", "date_of_birth" : "2000-01-02"}
{"id" : "1", "date_of_birth" : "2000-01-03"}
{"id" : "2", "date_of_birth" : "2000-01-04"}
{"id" : "3", "date_of_birth" : "2000-01-05"}
{"id" : "3", "date_of_birth" : "2000-01-06"}
```

And here my stdout result :
``` ruby
{
       "@timestamp" => 2016-12-08T22:03:47.078Z,
    "date_of_birth" => [
        [0] "2000-01-01",
        [1] "2000-01-02",
        [2] "2000-01-03"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "1",
             "tags" => [
        [0] "aggregate"
    ]
}
{
       "@timestamp" => 2016-12-08T22:03:47.081Z,
    "date_of_birth" => [
        [0] "2000-01-04"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "2",
             "tags" => [
        [0] "aggregate"
    ]
}
{
       "@timestamp" => 2016-12-08T22:03:55.533Z,
    "date_of_birth" => [
        [0] "2000-01-05",
        [1] "2000-01-06"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "3",
             "tags" => [
        [0] "aggregate"
    ]
}
```

**Comment by iti888** [comment]
I am using logstash-filter-aggregate (2.5.0), logstash-input-jdbc (4.1.3) and logstash 5.0.2

**Comment by iti888** [comment]
### Using tcp input. It works fine. But not jdbc input. 
**_To my observation, your aggregation filter, outputs (indexes) all inputs first and then ouput (index overwrite) the aggregation result.
For JDBC input, somehow the timeout did not kick in (or it did kick in since if I drop the non aggregated one by adding the following filter
    timeout_tags => ['aggregated']
           }
           if "aggregated" not in [tags] {
              drop {}
           }), the 3rd original doc did get dropped.
So the last doc was the original 3rd doc without aggregation. 
As for TCP input, the timeout DID kick in and the aggregation overwrite the last doc (3rd doc) and hence the result is correct._**:
### See the JDBC rubydebug output
{
       "@timestamp" => 2016-12-09T18:51:57.606Z,
    "date_of_birth" => 2000-01-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.637Z,
    "date_of_birth" => 2000-02-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.646Z,
    "date_of_birth" => 2000-03-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.648Z,
    "date_of_birth" => 2000-04-01T08:00:00.000Z,
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.648Z,
    "date_of_birth" => 2000-05-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.649Z,
    "date_of_birth" => 2000-06-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.762Z,
    "date_of_birth" => [
        [0] 2000-01-01T08:00:00.000Z,
        [1] 2000-02-01T08:00:00.000Z,
        [2] 2000-03-01T08:00:00.000Z
    ],
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T18:51:57.763Z,
    "date_of_birth" => [
        [0] 2000-04-01T08:00:00.000Z
    ],
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => []
}

### Here is the TCP rubydebug output

{
       "@timestamp" => 2016-12-09T19:04:20.218Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-01",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "1",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.220Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-02",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "1",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.221Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-03",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "1",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.221Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-04",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "2",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.223Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-05",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "3",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.224Z,
             "port" => 45950,
    "date_of_birth" => "2000-01-06",
         "@version" => "1",
             "host" => "127.0.0.1",
               "id" => "3",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.384Z,
    "date_of_birth" => [
        [0] "2000-01-01",
        [1] "2000-01-02",
        [2] "2000-01-03"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "1",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:20.388Z,
    "date_of_birth" => [
        [0] "2000-01-04"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "2",
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T19:04:30.372Z,
    "date_of_birth" => [
        [0] "2000-01-05",
        [1] "2000-01-06"
    ],
             "name" => nil,
         "@version" => "1",
      "description" => nil,
               "id" => "3",
             "tags" => []
}
### Only difference is that JDBC missing the last aggregation DOC.

**Comment by fbaligand** [comment]
Ok. 
I tried also with logstash 5.0.2 and it worked fine for me. 
Can you try to use timeout_tags option to clearly distinct aggregated events from original jdbc line events ?

Fabien

**Comment by iti888** [comment]
### Did you use Oracle to test? TCP input has no problem.
**_Here is the rubydebug after adding  timeout_tags => ['aggregated']_**

{
       "@timestamp" => 2016-12-09T22:38:53.917Z,
    "date_of_birth" => 2000-06-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.779Z,
    "date_of_birth" => 2000-01-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.888Z,
    "date_of_birth" => 2000-02-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.891Z,
    "date_of_birth" => 2000-03-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.892Z,
    "date_of_birth" => 2000-04-01T08:00:00.000Z,
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.896Z,
    "date_of_birth" => 2000-05-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-09T22:38:53.929Z,
    "date_of_birth" => [
        [0] 2000-01-01T08:00:00.000Z,
        [1] 2000-02-01T08:00:00.000Z,
        [2] 2000-03-01T08:00:00.000Z
    ],
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => [
        [0] "aggregated"
    ]
}
{
       "@timestamp" => 2016-12-09T22:38:53.931Z,
    "date_of_birth" => [
        [0] 2000-04-01T08:00:00.000Z
    ],
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => [
        [0] "aggregated"
    ]
}

**Comment by iti888** [comment]
BTW, I have upgrade ES to 5.1.1 LS to 5.1.1 etc. Same result.

**Comment by fbaligand** [comment]
I think I know what is your problem !
Do you set "-w 1" option at logstash startup ?

If you don't do that, you have no warranty that logs are processed in the right order. 
And this is necessary for aggregate plugin to work fine. 

Fabien

**Comment by iti888** [comment]
I did add -w 1 at the command line 
sudo ./bin/logstash -w 1 -f oracle-out_test.conf --path.settings=/etc/logstash

and at the /etc/logstash/logstash.yml
pipeline.workers: 1

Still the same result. Is that the JDBC input won't timeout or your filter does not know JDBC is done?

**Comment by fbaligand** [comment]
I think I found your problem !
In your sql query, you have to add "order by id "

Fabien

**Comment by iti888** [comment]
### Same result. It is not the order issue here.
**_I think for JDBC input, logstash does not know EOF of a query. Don't know if it is logstash bug or logstash jdbc input plugin bug or your aggregation filter bug.
Basically can not tell if the input stream has ended._**

[2016-12-09T18:38:10,021][INFO ][logstash.inputs.jdbc     ] (0.008000s) select t1.id,t1.name,t1.description, t2.date_of_birth from t1 inner join t2 on t1.ID=t2.T1_ID order by 1
[2016-12-09T18:38:10,441][INFO ][logstash.pipeline        ] Starting pipeline {"id"=>"main", "pipeline.workers"=>1, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>5, "pipeline.max_inflight"=>125}

{
       "@timestamp" => 2016-12-10T02:38:10.024Z,
    "date_of_birth" => 2000-01-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.460Z,
    "date_of_birth" => 2000-02-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.461Z,
    "date_of_birth" => 2000-03-01T08:00:00.000Z,
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.466Z,
    "date_of_birth" => 2000-04-01T08:00:00.000Z,
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.467Z,
    "date_of_birth" => 2000-05-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.468Z,
    "date_of_birth" => 2000-06-01T07:00:00.000Z,
             "name" => "name3",
         "@version" => "1",
      "description" => "3rd name",
               "id" => 3,
             "tags" => []
}
{
       "@timestamp" => 2016-12-10T02:38:10.486Z,
    "date_of_birth" => [
        [0] 2000-01-01T08:00:00.000Z,
        [1] 2000-02-01T08:00:00.000Z,
        [2] 2000-03-01T08:00:00.000Z
    ],
             "name" => "name1",
         "@version" => "1",
      "description" => "1st name",
               "id" => 1,
             "tags" => [
        [0] "aggregated"
    ]
}
{
       "@timestamp" => 2016-12-10T02:38:10.488Z,
    "date_of_birth" => [
        [0] 2000-04-01T08:00:00.000Z
    ],
             "name" => "name2",
         "@version" => "1",
      "description" => "2nd name",
               "id" => 2,
             "tags" => [
        [0] "aggregated"
    ]
}

**Comment by fbaligand** [comment]
Order is very important so that aggregate plugin works fine. 
In your previous output, the events ids were not in the right order. 
Now, in your last output, it is in the right order. 
However, I'm surprised that in your logs, it is written "order by 1". The right order is "order by t1.id"

Then, globally filter plugins are not dépendant from input plugins. It receives and processus json events whatever its source. 

Endly, the last aggregated event should append after 5 or 10 seconds (there is a flush every 5 seconds initiated by Logstash). 

Does your Logstash stop after the end of the sql query ?

Fabien

**Comment by iti888** [comment]
"order by 1" is the same as "order by t1.id". "1" means column position 1 in the query. If you are lazy to specify the column, you can just write it like this. "order by 1,3,5" that is order by col1,col3,col5.

I did not stop logstash.  But nothing happened after that. 

When I first discover this issue, it is when I am using only 1 record to see how aggregation works. It DOES nothing.(It showed me the original doc in ES. I did not use rubydebug output) When I used 2, then it showed me 1 record, then 2 when 3. That is why order is not the issue here. When there is only 1 record, order is not important.

**Comment by fbaligand** [comment]
When I speak about order, the very important thing so that it works fine, is that all events with id 2 have to come together.
If you have first an event with id 3, then an event with id 1, and then an event with id 3, aggregation will not work.
And in the previous output you provided, it was like that.

That said, unfortunately, the "order by" doesn't solve your problem...
Really surprising to me.
For information, which OS do you use ?

**Comment by iti888** [comment]
Ubuntu 16.0.4 desktop on VirtualBox running on windows 7 pro.

Sent from Yahoo Mail on Android 
 
  On Sat, Dec 10, 2016 at 10:20, Fabien Baligand<notifications@github.com> wrote:   
When I speak about order, the very important thing so that it works fine, is that all events with id 2 have to come together.
If you have first an event with id 3, then an event with id 1, and then an event with id 3, aggregation will not work.
And in the previous output you provided, it was like that.

That said, unfortunately, the "order by" doesn't solve your problem...
Really surprising to me.
For information, which OS do you use ?

**Comment by iti888** [comment]
@fbaligand 
Could you test with a database?

**Comment by iti888** [comment]
@fbaligand 
One more question:
Why is there a "tags" => [[0] "aggregated"] on every aggregation? I thought only when timeout, then the tag was set.

**Comment by fbaligand** [comment]
I understand you can be surprised. But this is the intended behaviour.
Each time aggregate plugin detects an event with new id (different than the one in last aggregate map), it considers the last aggregate map as in timeout.
So each option about timeout is applied (tags, code, task_id_field).

This is because the standard aggregate use is to have a specific event which explicitly ends the aggregation with `end_of_task => true`.

**Comment by iti888** [comment]
@fbaligand 
The why the 3rd doc does not have a timeout event?

I have added
	push_map_as_event_on_timeout => true
 	timeout_code => "event.set('id1', event.get('id'))"
	timeout_task_id_field => "name"

And rubydebug output looks extact like I have changed nothing. I would expect something extra

Looks like the filter is still waiting for an event and did not timeout.

**Comment by iti888** [comment]
Interesting:
I added in input:
    tcp {
    		port => 12345
    		codec => json
	}
as input along with JDBC input.  Now the 3rd doc appears. I DID not enter anything though port 12345.

What was this? Logstash bug?

**Comment by fbaligand** [comment]
Sincerely, I have no idea why it does not generate a third "aggregated"
timeout event with jdbc input.
There's no reason to have a different behaviour depending the input.

Maybe it is logstash bug, maybe a jdbc input bug.
Anyway, that is very surprising and weird...

2016-12-12 23:37 GMT+01:00 iti888 <notifications@github.com>:

**Comment by fbaligand** [comment]
This is really really surprising behaviour !
Definitively, the issue is not caused by aggregate plugin itself.
Clearly, I see there a Logstash core bug or a jdbc input plugin bug, and I have sincerely no explanation for this behaviour.
I invite you to open an issue in https://github.com/elastic/logstash

Anyway, I'm happy for you : using this trick, your use case finally works !

**Comment by fbaligand** [comment]
@iti888 
Now you finally solved your problem (and despite it is weird), can I close the issue ?

**Comment by iti888** [comment]
I am going to test with mutiple config files. Have to wait and see if the workaround works for that.

**Comment by fbaligand** [comment]
@iti888 
Fine. Keep me informed.

**Comment by iti888** [comment]
@fbaligand 
Multiple files still works with the workaround. 
But I just test the same config without aggregation filter and saw the last record show up in Rubydebug.
This confirm the bug is in your aggregation filter. Since when there is no filter, the last record show, but when aggregation filter is used, the last record did not show. 
Thanks

**Comment by fbaligand** [comment]
Indeed, that's intriguing...
I have to reproduce it.

**Comment by Cinerar** [comment]
not sure, but this could be similar to this issue: 
if i use file input there should be new line at the end of the file or last entry will be ignored.

**Comment by iti888** [comment]
@Cinerar 
Yeah, I think there is an issue with the filter handling the last event. @fbaligand should be able to figure this out now since your case is easier to reproduce.

**Comment by fbaligand** [comment]
I just tested @iti888 use case described in this issue (with tables t1 and t2).

So :
- The good news is that I reproduce your problem : the last 'aggregated' event (id=3) is not produced.
- But the explanation is quite simple : when you don't set a jdbc input `schedule` option, as soon as all generated events are processed, Logstash is stopped. So aggregate plugin's timeout mechanism doesn't have time to happen.
- If you set tcp input plugin, then logstash is not stopped after all result rows are processed, so aggregate timeout happens, and the "aggregated" event with id=3 is produced.

Compared to your initial comment, things I have changed in my test configuration :
- add `-w 1 ` flag at logstash startup
- add ` order by t1.id` at end of jdbc input `statement` option
- replace `map['date_of_birth'] <<= event.get('date_of_birth')` by `map['date_of_birth'] << event.get('date_of_birth')` 
- set `stdout` as output plugin (instead of elasticsearch)

**Comment by iti888** [comment]
If this is the case, other than those input without ending, by using the aggregate filter, all will have a problem with missing last record. But @Cinerar's case with file input has no problem with newline at the end of file, how do you explain that? Last line with or without NL should terminate Logstash and hence no result for both. But that is not the case.

**Comment by fbaligand** [comment]
OK, so :
- if file input don't see a new line at end-of-file, the last line is not processed by file input. You can set whatever filter (or no filter at all), it will not change : that is the intended file input behaviour. I invite you to check for yourself
- so last line without NL in file input has no relation with aggregate filter
- when file input finishes to read a file, it doesn't stop Logstash
- to my knowledge, only jdbc input plugin has such a behaviour

**Comment by fbaligand** [comment]
That said, concerning your issue, now I reproduce and understand your problem, I made tests to try to fix aggregate plugin behaviour in that case.
And good news : I managed to fix the issue !
So now I'll finalize the code, make unit tests, and will soon release a new version with this fix.

**Comment by iti888** [comment]
Thanks a lot for working through the holidays.Happy New Year !

Sent from Yahoo Mail on Android 
 
  On Sat, Dec 31, 2016 at 0:53, Fabien Baligand<notifications@github.com> wrote:   
That said, concerning your issue, now I reproduce and understand your problem, I made tests to try to fix aggregate plugin behaviour in that case.
And good news : I managed to fix the issue !
So now I'll finalize the code, make unit tests, and will soon release a new version with this fix.

**Comment by fbaligand** [comment]
Happy new year ! :)

Fabien

**Comment by iti888** [comment]
@fbaligand 
Looks like you can detect the last record of the query. Could you add a column in the filter to signify the last record of the query? I need this capability, but have no idea how to get it.
Thanks

**Comment by fbaligand** [comment]
To be precise, aggregate plugin can detect last flush just before Logstash shutdown, which is quite different.

**Comment by iti888** [comment]
For each input steam, looks like the plugin uses the timeout to detect the last record or true timeout. Can you add a field to signify the last aggregation?

**Comment by iti888** [comment]
Never mind, I was able to detect the last record using timeout_task_id_field. When this field is the same as task_id, it is the last record.

**Comment by fbaligand** [comment]
@iti888 
Actually, the fact that timeout_task_id_field value is different from you real task_id is a bug...
I detected it last week, and the bug will be solved in next version.

Anyway, I don't consider issue as solved : if you don't define schedule option (one run and then Logstash shutdown), it have to flush the last aggregate map.
As mentioned above, this will be released soon.

**Comment by iti888** [comment]
@fbaligand 
So, you fixed a bug and I will have to find other way to detect the last record? Can you keep this bug? Or add another field to keep this function?
BTW, I do use schedule="* * * * *". Basically run every minutes for all the input.

**Comment by fbaligand** [comment]
Question : when you use schedule option, you don't have the bug you described in this issue description ? I mean : the last sql result is lost.

**Comment by iti888** [comment]
No. Let me describe how it is used.
aggregate {
task_id => "%{id}"
....
timeout_task_id_field => "last_record_id"}

assume you have record with id: 1,1,2,2,2,6
in output:
last_record_id will have the following value when each event occur:
1,2,2,2,6,6 (basically  last_record_id is next event's task_id, except last record is its own task_id)
when 6=6, it is the last record.
If you fix the bug and the last_record_id become:
1,1,2,2,2,6
then I can not detect last record anymore.
Schedule or not does not effect the above result.
So, please keep this feature:)

**Comment by fbaligand** [comment]
Well, it's not a feature, but a bug :)
And I can't keep - voluntary - a bug. Thus it's already fixed.
Then, you do not have to update your aggregate plugin version and keep this "bug-feature".

That said, we can find another way for you to detect last event.
If you don't use "schedule" feature, I can add a special tag "_aggregatefinalflush" when the special final flush just before Logstash shutdown happens.

And If you use "schedule", you can store in map, the first event date, and then use timeout_code option to compute time difference between map start and end. If it is superior to timeout time, then it is last line.

Example : 
```
	aggregate {
		task_id => "%{id}"
		code => "
			map['@timestamp'] ||= event.timestamp.time
			map['id'] ||= event.get('id')
                        ...
		"
		push_previous_map_as_event => true
		timeout => 3
		timeout_code => "
			timeout = Time.now - event.timestamp.time
			event.tag('last') if timeout >= 3
		"
	}	
```

**Comment by iti888** [comment]
I can not use the  outdated plugin, you may never know if there is anything that will force me to upgrade and the "feature" will be lost:)
I will use the way you suggested. A lot more coding:) Was just one line...

When will the new plugin coming out?

**Comment by fbaligand** [comment]
Hi @iti888,

I have great news for you !
I just released aggregate v2.5.1 !

It includes especially :
- the last aggregate map is flushed, just before Logstash shutdown
- `aggregatefinalflush` tag is added in final flush, just before Logstash shutdown
- task_id_field feature is fixed when push_previous_map_as_event is enabled

Feel free to test this version and give me feedback !

**Comment by iti888** [comment]
Just tested with 2.5.1
Here are the observations:
1. @timestamp field is auto-added in map, which I think is not necessary if implemented the current way. The flush event's @timestamp is set at the timeout time not the event start time. That is event.timestamp.time +timeout. Is this your intent or a bug?
2. task_id_field does display the current event field value instead of next event field, which breaks my last event detection:)
So, now the actual code to implement the last record detection is:

aggregate {
		...
		code => "
			map['@timestamp1'] ||= Time.now
                        ...
		"
		push_previous_map_as_event => true
		timeout => 3
		timeout_code => "
	     	          elapsed_time =  Time.now.to_i - event.get('@timestamp1').to_i
		          event.tag('last1') if elapsed_time >= 3
		"
	}	
Direct using Time.now arithmetic will get error: TypeError: can't convert Timestamp into Rational

**Comment by fbaligand** [comment]
Concerning "task_id_field" feature, you were warned :)

Concerning timestamp :
- First, if don't write the same code I suggested, don't be surprised that the behavior is not the same :)
- Then, it's not a aggregate bug/feature, but à Logstash core feature. There is no auto generated @timestamp in aggregate map. But when you create a new event from aggregate map, if Logstash does not see @timestamp field in map, it generates @timestamp field using Time.now. So when a timeout new event is generated, indeed, the timestamp is equal to timeout time. 

Knowing that, my suggested code uses initial event timestamp to compute elapsed time, and fills timeout event with initial event timestamp.

**Comment by iti888** [comment]
@fbaligand  The code works either way.
Leaving @timestamp alone has some purpose, you can see if the timeout works:)
 Thanks for the help

**Comment by fbaligand** [comment]
OK. The most important is that it fulfills your needs !
Nice to see we finally solve all your problems !

**Comment by manivannansp** [comment]
@fbaligand  Looks like this issue started coming back after 2.9.1 version. while I am using Logstash7.8.0 i noticed this issue.

Version - 7.8 (2.9.1)
Look at the logs generated for the last record
{
      "@timestamp" => 2020-10-29T20:24:52.722Z,
              "id" => 89744628,
       "key" => "000000000004400748",
    "displayorder" => 1.0,
        "@version" => "1",
             "part" => "000000000004400748"
}
..................................................................................................................................................................[2020-10-29T13:24:54,471][INFO ][logstash.runner          ] Logstash shut down.
Completed processing


Version - Logstash 6.5 (2.6.0)

{
                "id" => 89744628,
          "@version" => "1",
    "productrecords" => [
        [0] {
            "key" => "000000000002989950",
                  "part" => "000000000002989950"
        },
        [1] {
            "key" => "000000000004869588",
                  "part" => "000000000004869588"
        },
        [2] {
            "key" => "000000000002989942",
                  "part" => "000000000002989942"
        },
        [3] {
            "key" => "000000000004400703",
                  "part" => "000000000004400703"
        },
        [4] {
            "key" => "000000000004400748",
                  "part" => "000000000004400748"
        }
    ],
        "@timestamp" => 2020-10-29T20:27:15.963Z,
              "tags" => [
        [0] "aggregated",
        [1] "_aggregatefinalflush"
    ]
}
.[2020-10-29T13:27:16,005][INFO ][logstas

**Comment by fbaligand** [comment]
In your tests, I see that you upgrade both plugin version and logstash version.  Could you try aggregate version 2.9.1 on logstash 6.5 ?

There are some internal changes in logstash 7.x that can be the reason of your problem.

Especially pipeline.java_execution=true
But also pipeline.ordered=false

You can try on Logstash 7 : 
pipeline.java_execution=false
Or pipeline.ordered=true

**Comment by manivannansp** [comment]
Thanks for the feedback @fbaligand  elastic confirmed it is a issue in logstash version, they fixed it in the next version. after upgrading 7.9.3 its working now

**Comment by fbaligand** [comment]
Thanks for this information.
It's interesting to know.

**Comment by robbavey** [comment]
@manivannansp We discovered an issue in versions `7.7.0` onwards where flushing of events was not working correctly with the `pipeline.ordered` flag. This has been fixed in versions `7.9.1` and above. Please retry using the latest logstash, and feel free to reopen this issue, or create a new one if you are still having issues.
